### PR TITLE
Allow mount.ceph to get mount info from ceph configs and keyrings

### PR DESCRIPTION
**Pull request by jtlayton**
This patchset is a fairly significant overhaul of the mount.ceph helper. The main change is new functionality that allows it to get monitor addresses from a local ceph.conf file (or DNS SRV records) and cephx secrets from local keyring files. A new mount option (conf=) is also added that allows the admin to direct it to use a particular config file.

Since this program is almost always invoked with elevated privileges, it will spawn a child process that drops privileges to handle the config file parsing and pass back any info it finds to the parent via anonymous shared memory.

The config file is only consulted if we're missing either a mount secret value or monitor addresses.

### CODE REVIEW

**Comment by jtlayton** [comment]
This all seems to work rather well. The only real concern I have is that the ceph libraries tend to be fairly chatty on stderr, particularly when the config file can't be found.

Most people that are successfully the kernel client today will never notice that though since they'll usually specify the necessary info to avoid consulting the config file.  The main exception here is people who are disabling cephx. Those folks will generally not specify a secret, and so they may end up with the program trying to find a keyring even when it isn't necessary.

If we're concerned with that use-case, we could add some mount option to cue the helper not to search for a secret at all. Maybe "-o secret=<none>" since we're already handling "-o secret=<hidden>" as it is?

File: doc/cephfs/kernel.rst [hunk 1]
```diff
@@ -9,16 +9,20 @@ monitor host name(s) into IP address(es) for you. For example::
 	sudo mkdir /mnt/mycephfs
 	sudo mount -t ceph 192.168.0.1:6789:/ /mnt/mycephfs
 
-To mount the Ceph file system with ``cephx`` authentication enabled, you must
-specify a user name and a secret. ::
+To mount the Ceph file system with ``cephx`` authentication enabled, the kernel
+must authenticate with the cluster. The default ``name=`` option is ``guest``.
+The secret can be specified with the secret= option. ::
```
Comment (jtlayton) [inline doc/cephfs/kernel.rst]: Should probably have backticks around secret= in the above line. Will fix.

**Comment by jtlayton** [comment]
Weird. It says some SoB lines are missing, but they all seem to be there.

**Comment by batrick** [comment]
> Weird. It says some SoB lines are missing, but they all seem to be there.

The bot appears to be broken I think.

File: install-deps.sh [hunk 2]
```diff
@@ -342,7 +342,7 @@ else
 	$SUDO env DEBIAN_FRONTEND=noninteractive mk-build-deps --install --remove --tool="apt-get -y --no-install-recommends $backports" $control || exit 1
 	$SUDO env DEBIAN_FRONTEND=noninteractive apt-get -y remove ceph-build-deps
 	if [ "$control" != "debian/control" ] ; then rm $control; fi
-	$SUDO apt-get install -y libxmlsec1 libxmlsec1-nss libxmlsec1-openssl libxmlsec1-dev
+	$SUDO apt-get install -y libxmlsec1 libxmlsec1-nss libxmlsec1-openssl libxmlsec1-dev libcap-ng-dev
```
Comment (tchaikov) [inline install-deps.sh]: no need to install `libcap-ng-dev` separately. `debian/control` should already take care of this.

these dependencies are only for building the [python-saml](https://github.com/onelogin/python-saml) python extension. the SAML feature is not enabled by default.

File: install-deps.sh [hunk 3]
```diff
@@ -404,15 +404,15 @@ else
         IGNORE_YUM_BUILDEP_ERRORS="ValueError: SELinux policy is not managed or store cannot be accessed."
         sed "/$IGNORE_YUM_BUILDEP_ERRORS/d" $DIR/yum-builddep.out | grep -qi "error:" && exit 1
         # for building python-saml and its dependencies
-        $SUDO $yumdnf install -y xmlsec1 xmlsec1-nss xmlsec1-openssl xmlsec1-devel xmlsec1-openssl-devel libtool-ltdl-devel
+        $SUDO $yumdnf install -y xmlsec1 xmlsec1-nss xmlsec1-openssl xmlsec1-devel xmlsec1-openssl-devel libtool-ltdl-devel libcap-ng-devel
```
Comment (tchaikov) [inline install-deps.sh]: ditto.

File: install-deps.sh [hunk 4]
```diff
@@ -404,15 +404,15 @@ else
...
         ;;
     opensuse*|suse|sles)
         echo "Using zypper to install dependencies"
         zypp_install="zypper --gpg-auto-import-keys --non-interactive install --no-recommends"
         $SUDO $zypp_install systemd-rpm-macros
         munge_ceph_spec_in $with_seastar $for_make_check $DIR/ceph.spec
         $SUDO $zypp_install $(rpmspec -q --buildrequires $DIR/ceph.spec) || exit 1
-        $SUDO $zypp_install libxmlsec1-1 libxmlsec1-nss1 libxmlsec1-openssl1 xmlsec1-devel xmlsec1-openssl-devel
+        $SUDO $zypp_install libxmlsec1-1 libxmlsec1-nss1 libxmlsec1-openssl1 xmlsec1-devel xmlsec1-openssl-devel libcap-ng-devel
```
Comment (tchaikov) [inline install-deps.sh]: ditto.

File: src/mount/conf.cc [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+  conf.parse_env(cct->get_module_type()); // environment variables override
+  conf.apply_changes(nullptr);
+
+  MonClient monc = MonClient(cct.get());
+  err = monc.build_initial_monmap();
+  if (err)
+    goto scrape_keyring;
+
+  for (auto p = monc.monmap.addr_mons.begin(); p != monc.monmap.addr_mons.end(); ++p) {
```
Comment (tchaikov) [inline src/mount/conf.cc]: could use range-based loop, like
```c++
for (const auto& mon : monc.monmap.addr_mons) {
  auto& eaddr = p.first;
  // ...
}
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -395,6 +395,11 @@ if(WITH_LIBCEPHFS OR WITH_KRBD)
   find_package(keyutils REQUIRED)
 endif()
 
+if(WITH_LIBCEPHFS)
+  find_package(PkgConfig REQUIRED)
```
Comment (tchaikov) [inline CMakeLists.txt]: nit, could add `QUIET` before `REQUIRED`. as what we are interested in is libcap-ng. like
```cmake
find_package(PkgConfig QUIET REQUIRED)
```

also, i'd suggest move this check into https://github.com/ceph/ceph/blob/e909031d3c231052fdf0afe993f669514faadde6/src/CMakeLists.txt#L641 . i.e., closer to where it is used.

File: src/mount/mount.ceph.c [hunk 7]
```diff
@@ -172,53 +168,36 @@ static int parse_options(const char *data, struct ceph_mount_info *cmi)
...
-			   recognize our keytype; hence, it is skipped
-			   here and appended to options on add_key
-			   failure */
-			size_t len = sizeof(secret);
-			strncpy(secret, value, len-1);
-			secret[len-1] = '\0';
-			saw_secret = secret;
-			skip = true;
+			len = strnlen(value, sizeof(cmi->cmi_secret) + 1);
+			if (len <= sizeof(cmi->cmi_secret))
```
[the remark this answers is not in the document]
Comment (jtlayton) [reply src/mount/mount.ceph.c]: > i think we should error out if `len == sizeof(cmi->cmi_secret)`. in that case, we cannot store the terminating null.
> 

Good catch. I think the better fix is to remove the NULL from MAX_SECRET_LEN, and adjust the usage of it accordingly. I'll add a patch to do that.

> also, shall we return `-EINVAL` if `(len == sizeof(cmi->cmi_secret))`, something like

No. If the secret looks bogus, then we just want to ignore it and move on. If we hit that and the secret turned out not to be necessary for a successful mount, that could break an installation that was previously working. Better to just feed a blank secret to the kernel and let it sort it out. We could spew a warning though -- I'll think about that.

File: src/mount/mount.ceph.c [hunk 8]
```diff
@@ -172,53 +168,36 @@ static int parse_options(const char *data, struct ceph_mount_info *cmi)
...
-			   backwards compatibility, if add_key doesn't
-			   recognize our keytype; hence, it is skipped
-			   here and appended to options on add_key
-			   failure */
-			size_t len = sizeof(secret);
-			strncpy(secret, value, len-1);
-			secret[len-1] = '\0';
-			saw_secret = secret;
-			skip = true;
+			len = strnlen(value, sizeof(cmi->cmi_secret) + 1);
```
[the remark this answers is not in the document]
Comment (jtlayton) [reply src/mount/mount.ceph.c]: Another good catch. It's OK to pass a length that includes the NULL to strnlen, so probably what we want to do is just move the +1 outside the strnlen:
```
len = strnlen(value, sizeof(cmi->cmi_secret)) + 1;
```

File: src/mount/mount.ceph.c [hunk 9]
```diff
@@ -347,21 +348,27 @@ int main(int argc, char *argv[])
...
 		exit(EX_USAGE);
 	}
 
 	/* Ensure the ceph key_type is available */
 	modprobe();
 
+	rsrc = finalize_src(&cmi);
+	if (!rsrc) {
+		fprintf(stderr, "failed to resolve source\n");
+		exit(EX_USAGE);
```
[the remark this answers is not in the document]
Comment (jtlayton) [reply src/mount/mount.ceph.c]: Yeah, wouldn't hurt to do that. I'll fix that up.

File: src/mount/conf.cc (hunk 5 above)
Comment (jtlayton) [reply src/mount/conf.cc]: What's the benefit of doing so?

Comment (tchaikov) [reply src/mount/conf.cc]: i think the benefit is better readability and less repeating.

File: src/mount/mount.ceph.c [hunk 10]
```diff
@@ -96,6 +99,104 @@ static char *finalize_src(struct ceph_mount_info *cmi)
...
+	pid_t pid;
+	struct ceph_config_info *cci;
+
+	/* Don't do anything if we already have requisite info */
+	if (cmi->cmi_secret[0] && cmi->cmi_mons)
+		return 0;
+
+	cci = mmap((void *)0, sizeof(*cci), PROT_READ | PROT_WRITE,
+		   MAP_ANONYMOUS | MAP_SHARED, -1, 0);
+	if (cci == (struct ceph_config_info *)-1) {
```
Comment (luis-henrix) [inline src/mount/mount.ceph.c]: nit: Why not using`if (cci == MAP_FAILED)` instead?

File: src/mount/mount.ceph.c [hunk 11]
```diff
@@ -96,6 +99,104 @@ static char *finalize_src(struct ceph_mount_info *cmi)
...
+		mount_ceph_debug("Unable to allocate memory: %s\n",
+				 strerror(errno));
+		return EX_SYSERR;
+	}
+	memset(cci, '\0', sizeof(*cci));
```
Comment (luis-henrix) [inline src/mount/mount.ceph.c]: nit2: not sure this is really necessary, as MAP_ANONYMOUS is supposed to return a zero initialized mapping.

**Comment by luis-henrix** [comment]
In general, this looks good to me; I couldn't find any real issue in these changes (apart from the 2 minor comments).  I haven't tested this code, though.  I'll try to do that later today (or in the next few days).

File: src/mount/mount.ceph.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (jtlayton) [reply src/mount/mount.ceph.c]: I ended up doing a rework of the defines for this. I'm hoping the result is clearer now.

File: src/mount/mount.ceph.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (jtlayton) [reply src/mount/mount.ceph.c]: Ended up reworking the defines for this. Hopefully the result is clearer now.

File: src/mount/conf.cc (hunk 5 above)
Comment (jtlayton) [reply src/mount/conf.cc]: Fair enough. Changed to range-based loop.

**Comment by jtlayton** [comment]
I went ahead and folded the changes into the series, and force pushed the results.

File: src/mount/conf.cc [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+#include "common/ceph_context.h"
+#include "common/ceph_argparse.h"
+#include "global/global_init.h"
+#include "common/config.h"
+#include "auth/KeyRing.h"
+#include "mount.ceph.h"
+#include "mon/MonClient.h"
+
+#include <iostream>
```
Comment (tchaikov) [inline src/mount/conf.cc]: i assume this include was added for debugging, right?

Comment (jtlayton) [reply src/mount/conf.cc]: Yes, leftover from some cout << debugging that I was doing early on. Removed.

**Comment by batrick** [comment]
retest this please

File: doc/cephfs/kernel.rst [hunk 13]
```diff
@@ -9,16 +9,20 @@ monitor host name(s) into IP address(es) for you. For example::
...
 	sudo mount -t ceph 192.168.0.1:6789:/ /mnt/mycephfs -o name=admin,secret=AQATSKdNGBnwLhAAnNDKnH65FmVKpXZJVasUeQ==
 
 The foregoing usage leaves the secret in the Bash history. A more secure
 approach reads the secret from a file. For example::
 
 	sudo mount -t ceph 192.168.0.1:6789:/ /mnt/mycephfs -o name=admin,secretfile=/etc/ceph/admin.secret
-	
+
+If a secret is not specified in the options, then the mount helper will attempt
+to find it in a ceph keyring file.
```
Comment (batrick) [inline doc/cephfs/kernel.rst]: I think we could do a better job suggesting this new way is the preferred approach. Maybe reorder and suggest that secret= secrefile= are for finer control or mounts without a keyring file.

Comment (jtlayton) [reply doc/cephfs/kernel.rst]: Good point. I cleaned the document up a bit. Hopefully this will be a bit clearer.

File: src/mount/conf.cc [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+#include "mon/MonClient.h"
+
+extern "C" void mount_ceph_get_config_info(const char *config_file,
+					   const char *name,
+					   struct ceph_config_info *cci)
+{
+  int err;
+  KeyRing keyring;
+  CryptoKey secret;
+  std::string secret_str;
```
Comment (fullerdj) [inline src/mount/conf.cc]: `#include <string>`

File: src/mount/conf.cc [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+  std::string monaddrs;
+  vector<const char *> args = { "--name", name };
```
Comment (fullerdj) [inline src/mount/conf.cc]: `#include <vector>`

File: src/mount/conf.cc [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    if (first)
+      first = false;
+    else
+      monaddrs += ",";
+
+    monaddrs += addr;
+  }
+
+  if (monaddrs.length())
+    strcpy(cci->cci_mons, monaddrs.c_str());
```
Comment (fullerdj) [inline src/mount/conf.cc]: `#include <cstring>`

File: src/mount/conf.cc [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+  conf.parse_env(cct->get_module_type()); // environment variables override
+  conf.apply_changes(nullptr);
+
+  MonClient monc = MonClient(cct.get());
+  err = monc.build_initial_monmap();
+  if (err)
+    goto scrape_keyring;
+
+  for (const auto& mon : monc.monmap.addr_mons) {
+    auto& eaddr = mon.first;
```
Comment (fullerdj) [inline src/mount/conf.cc]: Even though you've used auto here, mon here is of type std::map, which requires `#include <map>`.

File: src/mount/conf.cc (hunk 14 above)
Comment (jtlayton) [reply src/mount/conf.cc]: Thanks, fixed (along with all of the other missing includes). I guess some of the other headers that this was including included these files, which is why it still built without them, but it is better to include them explicitly.

**Comment by jtlayton** [comment]
Rebased to deal with merge conflicts.

**Comment by liewegas** [comment]
On Fri, 13 Sep 2019, jtlayton wrote:
> Rebased to deal with merge conflicts.

FWIW I've started putting my PendingReleaseNotes items in random positions so that every single PR that adds an item doesn't cause a conflict.

**Comment by jtlayton** [comment]
> On Fri, 13 Sep 2019, jtlayton wrote: Rebased to deal with merge conflicts.
> FWIW I've started putting my PendingReleaseNotes items in random positions so that every single PR that adds an item doesn't cause a conflict.

Good idea. I'll do that if I need to rebase again.

**Comment by jtlayton** [comment]
Retest this please

**Pull request merged**